src/systems/world-update.js

```diff
@@ -0,0 +1,192 @@
+          target = new THREE.Vector3();
+        }
+
+        this.updateMatrices();
+        this.matrixWorld.decompose(position, quaternion, target);
+
+        return target;
+      };
+    })();
+
+    THREE.Object3D.prototype.updateMatrix = function() {
+      this.matrix.compose(
+        this.position,
+        this.quaternion,
+        this.scale
+      );
+
+      this.matrixWorldNeedsUpdate = true;
+
+      if (!this.matrixIsModified) {
+        this.matrixIsModified = true;
+
+        if (this.cachedMatrixWorld) {
+          this.cachedMatrixWorld.copy(this.matrixWorld);
+          this.matrixWorld = this.cachedMatrixWorld;
+        }
+      }
+    };
+
+    THREE.Object3D.prototype.applyMatrix = function(matrix) {
+      this.matrix.multiplyMatrices(matrix, this.matrix);
+      this.matrix.decompose(this.position, this.quaternion, this.scale);
+
+      this.matrixWorldNeedsUpdate = true;
+
+      if (!this.matrixIsModified) {
+        this.matrixIsModified = true;
+
+        if (this.cachedMatrixWorld) {
+          this.cachedMatrixWorld.copy(this.matrixWorld);
+          this.matrixWorld = this.cachedMatrixWorld;
+        }
+      }
+    };
+
+    // By the end of this function this.matrix reflects the updated local matrix
+    // and this.matrixWorld reflects the updated world matrix, taking into account
+    // parent matrices.
+    //
+    // forceLocalUpdate - Forces the local matrix to be updated regardless of if it has not
+    // been marked dirty.
+    //
+    // forceWorldUpdate - Forces the world matrix to be updated regardless of if the local matrix
+    // has been updated since the last update.
+    //
+    // skipParents - unless true, all parent matricies are updated before updating this object's
+    // local and world matrix.
+    //
+    // Returns true if the world matrix was updated.
+    //
+    THREE.Object3D.prototype.updateMatrices = function(forceLocalUpdate, forceWorldUpdate, skipParents) {
+      if (!this.hasHadFirstMatrixUpdate) {
+        if (
+          !this.position.equals(zeroPos) ||
+          !this.quaternion.equals(zeroQuat) ||
+          !this.scale.equals(oneScale) ||
+          !this.matrix.equals(identity)
+        ) {
+          // Only update the matrix the first time if its non-identity, this way
+          // this.matrixIsModified will remain false until the default
+          // identity matrix is updated.
+          this.updateMatrix();
+        }
+
+        this.hasHadFirstMatrixUpdate = true;
+        this.cachedMatrixWorld = this.matrixWorld;
+      } else if (this.matrixNeedsUpdate || this.matrixAutoUpdate || forceLocalUpdate) {
+        this.updateMatrix();
+        if (this.matrixNeedsUpdate) this.matrixNeedsUpdate = false;
+      }
+
+      if (!skipParents && this.parent) {
+        this.parent.updateMatrices(false, forceWorldUpdate, false);
+      }
+
+      if (this.matrixWorldNeedsUpdate || forceWorldUpdate) {
```

Comment:
Don't we also need to do this if any parent matrix was updated above?

Reply:
that sounds more correct